src/idl_gen_python.cpp

```diff
@@ -1763,26 +1823,41 @@ class PythonGenerator : public BaseGenerator {
           // Scalar or sting fields.
           field_type = GetBasePythonTypeForScalarAndString(base_type);
           if (field.IsScalarOptional()) {
+            import_typing_list.insert("Optional");
             field_type = "Optional[" + field_type + "]";
           }
           break;
       }
 
       const auto default_value = GetDefaultValue(field);
-      // Wrties the init statement.
       const auto field_field = namer_.Field(field);
-      code += GenIndents(2) + "self." + field_field + " = " + default_value +
-              "  # type: " + field_type;
+
+      // If the default value is None, the type must be Optional, unless it's a
+      // Union that already contains None.
+      const bool is_already_optional =
+          (field_type.rfind("Optional[", 0) == 0) ||
+          (field_type.rfind("Union[", 0) == 0);
+      if (default_value == "None" && !is_already_optional) {
+        import_typing_list.insert("Optional");
+        field_type = "Optional[" + field_type + "]";
+      }
+
+      // Build signature with keyword arguments, type hints, and default values.
+      signature_params +=
+          ", " + field_field + ": " + field_type + " = " + default_value;
```

Comment:
Typing is opt-in with `--python-typing`, no typing imports or annotations should be output if not explicitly requested.

Reply:
Best I can tell when looking at the generator code, when `--python-typing` is specified it will output at least some typing annotations in the .py file.

Reply:
The generated results looks good to me, though it does conflict with some recent changes. (see #8623)

I'm not sure how often people look at .pyi files directly, but it might be useful to have one parameter per line for those as well to avoid the lines becoming super long.